jobs/every5minstat.js

```diff
@@ -1,55 +1,90 @@
-                                        }
-                                    });
-                                }
-                            });
-                        }
-                    });
-                }
-            });
+var cronJob = require('cron').CronJob,
+    logger = require('../logger'),
+    redis = require('../redis-helper'),
+    // Mongoose models
+    User = require('../models/user'),
+    Openhab = require('../models/openhab'),
+    UserDevice = require('../models/userdevice'),
+    Invitation = require('../models/invitation'),
+    Myohstat = require('../models/myohstat'),
+    stats = [],
+    expectedFinishCount, openhabCount, userCount, openhabOnlineCount, invitationUsedCount,
+    invitationUnusedCount, userDeviceCount;
+
+/**
+ * Callback function for a count operation on a Mongoose model. It will save the count in the
+ * stats object, if there was no error, and calls the checkFinishedAndSave function to save the
+ * statistics, if all counts are finished.
+ *
+ * Tis function expects, that the this argument is the name of the statistic value.
+ *
+ * @private
+ */
+function countCallback (err, count) {
+    if (!err) {
+        stats[this] = count;
+    }
+
+    checkFinishedAndSave();
+}
+
+/**
+ * Checks, if all callbacks are executed and if so, validates the stats object. It will then, if the
+ * validation succeeds, save the values to redis.
+ */
+function checkFinishedAndSave() {
+    var newStat = new Myohstat({uC: userCount, oC: openhabCount,
+        ooC: openhabOnlineCount, iuC: invitationUsedCount,
+        iuuC: invitationUnusedCount, udC: userDeviceCount});
+
+    expectedFinishCount--;
+    if (expectedFinishCount !== 0) {
+        // not finished, yet
+        return;
+    }
+
+    // validate the results
+    if (Object.keys(stats).length !== 6) {
+        // not all data could be retrieved
+        return;
+    }
+
+    newStat.save();
+    // Set current statistics to redis
+    redis.mset(
+        [
+            'openhabCount',
+            stats['openhabCount'],
+            'openhabOnlineCount',
+            stats['openhabOnlineCount'],
+            'userCount',
+            stats['userCount'],
+            'invitationUsedCount',
+            stats['invitationUsedCount'],
+            'invitationUnusedCount',
+            stats['invitationUnusedCount'],
+            'userDeviceCount',
+            stats['userDeviceCount'],
+            'last5MinStatTimestamp',
+            new Date()
+        ],
+        function (err, result) {
+            logger.info('openHAB-cloud: every5min statistics collection job finished');
         }
-    });
+    );
+}
+
+module.exports = new cronJob('00 */5 * * * *', function () {
+    logger.info('openHAB-cloud: every5min statistics collection job started');
+
+    // number of invocations of the finishSent() function to know, when all requests are finished to track multiple
+    // async actions
+    expectedFinishCount = 6;
```

Comment:
So i posted this 2 days ago, but did not click the "start review" button, that's twice that new feature from github has bit me.